integration/shareddata/benchmark_provider.go

```diff
@@ -0,0 +1,133 @@
+// Copyright 2021 FerretDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package shareddata
+
+import (
+	"crypto/sha256"
+	"encoding/hex"
+	"encoding/json"
+	"errors"
+	"reflect"
+
+	"go.mongodb.org/mongo-driver/bson"
+
+	"github.com/FerretDB/FerretDB/internal/util/iterator"
+	"github.com/FerretDB/FerretDB/internal/util/must"
+)
+
+// BenchmarkProvider is implemented by shared data sets that provide documents for benchmarks.
+type BenchmarkProvider interface {
+	// Name returns benchmark provider name.
+	Name() string
+
+	// NewIterator returns a new iterator for documents in the same order.
+	NewIterator() iterator.Interface[struct{}, bson.D]
+}
+
+// hashBenchmarkProvider checks that BenchmarkProvider's NewIterator methods returns a new iterator
+// for the same documents in the same order,
+// and returns a hash of those documents that could be used as a part of benchmark name.
+func hashBenchmarkProvider(bp BenchmarkProvider) string {
```

Comment:
So we moved from calculating hash of exact documents that are inserted to the collection, to calculating hash before inserting them. I see some potential disadvantages like for example the generator might be dependent on unix timestamp in seconds, so it might pass the `hashBenchmarkProvider` comparison but afterwards insert different documents, although it's really unusual situation, and this change will allow us to have a test name before inserting documents so I like it :)

Reply:
that's one of the reason we iterate twice and compare results